firedrake/interpolation.py

```diff
@@ -25,395 +27,303 @@
-            op2mat = self.callable()
-            petsc_mat = op2mat.handle
-            if needs_adjoint:
-                # Out-of-place Hermitian transpose
-                petsc_mat.hermitianTranspose(out=res)
-            elif tensor:
+            petsc_mat = result.handle
+            if tensor:
                 petsc_mat.copy(tensor.petscmat)
-            else:
-                res = petsc_mat
-            return tensor or firedrake.AssembledMatrix(arguments, self.bcs, res)
+                return tensor
+            return firedrake.AssembledMatrix(self.expr_args, bcs, petsc_mat)
         else:
-            # Assembling the action
-            cofunctions = ()
-            if needs_adjoint:
-                # The renumbered Interpolate has dropped Cofunctions.
-                # We need to explicitly operate on them.
-                dual_arg, _ = self.ufl_interpolate.argument_slots()
-                if not isinstance(dual_arg, ufl.Coargument):
-                    cofunctions = (dual_arg,)
-
-            if needs_adjoint and len(arguments) == 0:
-                Iu = self._interpolate(default_missing_val=default_missing_val)
-                return assemble(ufl.Action(*cofunctions, Iu), tensor=tensor)
-            else:
-                return self._interpolate(*cofunctions, output=tensor, adjoint=needs_adjoint,
-                                         default_missing_val=default_missing_val)
+            assert isinstance(tensor, Function | Cofunction | None)
+            return tensor.assign(result) if tensor else result
+
+
+def get_interpolator(expr: Interpolate) -> Interpolator:
+    """Create an Interpolator.
+
+    Parameters
+    ----------
+    expr : Interpolate
+        Symbolic interpolation expression.
+
+    Returns
+    -------
+    Interpolator
+        An appropriate :class:`Interpolator` subclass for the given
+        interpolation expression.
+    """
+    arguments = expr.arguments()
+    has_mixed_arguments = any(len(arg.function_space()) > 1 for arg in arguments)
+    if len(arguments) == 2 and has_mixed_arguments:
+        return MixedInterpolator(expr)
+
+    operand, = expr.ufl_operands
+    target_mesh = expr.target_space.mesh()
+    source_mesh = extract_unique_domain(operand) or target_mesh
+    submesh_interp_implemented = (
+        all(isinstance(m.topology, firedrake.mesh.MeshTopology) for m in [target_mesh, source_mesh])
+        and target_mesh.submesh_ancesters[-1] is source_mesh.submesh_ancesters[-1]
+        and target_mesh.topological_dimension == source_mesh.topological_dimension
+    )
+    if target_mesh is source_mesh or submesh_interp_implemented:
+        return SameMeshInterpolator(expr)
+
+    target_topology = target_mesh.topology
+    source_topology = source_mesh.topology
+
+    if isinstance(target_topology, VertexOnlyMeshTopology):
+        if isinstance(source_topology, VertexOnlyMeshTopology):
+            return VomOntoVomInterpolator(expr)
```

Comment:
Does `VomOntoVomInterpolator` support `MixedFunctionSpace`, or are we never expecting to build one for this case? There should be a `ValueError` in the constructor if we go for the latter.